src/Microsoft.OData.Core/UriParser/Resolver/CaseInsensitiveSchemaElementsCache.cs

```diff
@@ -0,0 +1,136 @@
+    /// Cache used to store schema elements using case-normalized names
+    /// to speed up case-insensitive model lookups. The cache is populated
+    /// up front so that if an item is not found in the cache, we can assume
+    /// it doesn't exist in the model. For this reason, it's important that
+    /// the model be immutable.
+    /// </summary>
+    internal sealed class CaseInsensitiveSchemaElementsCache
+    {
+        private static readonly List<IEdmSchemaElement> emptyList = new List<IEdmSchemaElement>();
+        // This cache is meant to be populate up front and remain read-only
+        // after that. Therefore, it doesn't need
+        // to be a concurrent dictionary.
+        private readonly Dictionary<string, List<IEdmSchemaElement>> cache;
+
+        /// <summary>
+        /// Builds a case-insensitive cache of schema elements from
+        /// the specified <paramref name="model"/>.
+        /// </summary>
+        /// <param name="model">The model whose schema elements to cache. This model should be immutable. See <see cref="ExtensionMethods.MarkAsImmutable(IEdmModel)"/>.</param>
+        public CaseInsensitiveSchemaElementsCache(IEdmModel model)
+        {
+            Dictionary<string, List<IEdmSchemaElement>> cache = new Dictionary<string, List<IEdmSchemaElement>>();
+
+            PopulateSchemaElements(model, cache);
+
+            foreach (IEdmModel referencedModel in model.ReferencedModels)
+            {
+                PopulateSchemaElements(referencedModel, cache);
+            }
+
+            this.cache = cache;
+        }
+
+        /// <summary>
+        /// Find all schema elements that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<IEdmSchemaElement> FindElements(string qualifiedName)
+        {
+            if (cache.TryGetValue(qualifiedName.ToUpperInvariant(), out List<IEdmSchemaElement> results))
+            {
+                return results;
+            }
+
+            return emptyList;
+        }
+
+        /// <summary>
+        /// Find all schema elements of type <typeparamref name="T"/> that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of schema elements to match.</typeparam>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements of type <typeparamref name="T"/> that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<T> FindElementsOfType<T>(string qualifiedName) where T : IEdmSchemaElement
+        {
+            IReadOnlyList<IEdmSchemaElement> elements = FindElements(qualifiedName);
+
+            List<T> results = new List<T>();

```

Comment:
I think in most cases it would not be larger. I assume for most schemas duplicate names will only exist for function overloads. 

Maybe I could also first check whether all the elements are of the same type, and if they are, then just return the original list without allocating a new one. This might require 2 loops, but if the list size is small, it may be worth it.

Reply:
I would not add 2 loops I think. 

Either keep the code as is or provide a capacity hint which in this case seems it should be 'elements.Count'. As long as the hint is not expected to be drastically different from the actual number of elements added in the end, providing a good capacity hint is meant to reduce internal allocations overhead to expand the list on insertions.

Reply:
Actually, I can't return an `IReadOnlyList<IEdmSchemaElements>` cast an `IReadOnlyList<T>` where `T` is a subtype, cause of covariance rules. So, if I wanted to completely eliminate this `List<T>` allocation, I would need cache the elements in different collections based on type. Maybe a cache of <string, IEdmSchemaType>, a cache of <string, IEdmOperation>, etc. Do you think that would be worthwhile?

Reply:
So, if I wanted to completely eliminate this `List<T>` allocation, I would need cache the elements in different collections based on type. Maybe a cache of <string, IEdmSchemaType>, a cache of <string, IEdmOperation>, etc. Do you think that would be worthwhile?

Reply:
The cost here is that we'd have multiple dictionaries and maybe that could have a larger overhead in some cases.